lib/services/local/saml_idp_service_provider.go

```diff
@@ -40,11 +50,24 @@ const (
 
 // SAMLIdPServiceProviderService manages IdP service providers in the Backend.
 type SAMLIdPServiceProviderService struct {
-	svc generic.Service[types.SAMLIdPServiceProvider]
+	svc        generic.Service[types.SAMLIdPServiceProvider]
+	log        logrus.FieldLogger
+	httpClient *http.Client
 }
 
 // NewSAMLIdPServiceProviderService creates a new SAMLIdPServiceProviderService.
-func NewSAMLIdPServiceProviderService(backend backend.Backend) (*SAMLIdPServiceProviderService, error) {
+func NewSAMLIdPServiceProviderService(backend backend.Backend, httpClient *http.Client) (*SAMLIdPServiceProviderService, error) {
```

Comment:
Instead of changing the function signature and having to pass `nil` everywhere, can we use functional arguments? So only callers that need to provide a custom http client will use it.

Reply:
@flyinghermit Sorry if I was unclear, you would still change the function signature to accept a list of functional options but clients that don't want to provide it won't need to do so. Something like this:

```go
type SAMLIdPOption func(*SAMLIdPServiceProviderService)

func WithHTTPClient(httpClient *http.Client) SAMLIdPOption{
  return func(s *SAMLIdPServiceProviderService) {
    s.httpClient = httpClient
  }
}

func NewSAMLIdPServiceProviderService(backend backend.Backend, opts ...SAMLIdPOption) xxx {
}

// The caller that doesn't need to set custom client, nothing changes:
NewSAMLIdPServiceProviderService(backend)

// The caller that needs to set custom httpClient:
NewSAMLIdPServiceProviderService(backend, WithHTTPClinet(httpClient))
```

Does that make sense?

Reply:
Oh got it. That would also make it easy to add/change new options in the future.

Reply:
added in https://github.com/gravitational/teleport/pull/35454/commits/8fa2923428f124ae405a84966e886658e10815d1. Thanks for flagging, its much cleaner now!